eth_tools/transfer_event_parser.py

```
@@ -0,0 +1,118 @@
+        self.end = end
+        self.keys = {}
+        self.keys['amount'] = 'amount'
+        self.keys['to'] = 'to'
+        self.keys['from'] = 'from'
+
+    def parse_events(self, events: list):
+        first_event = True
+        for event in events:
+            if event['event'] == 'Transfer':
+                if first_event:
+                    self.set_quantity_key(event)
+                    self.start = self.start if self.start is not None else event['blockNumber']
+                    first_event = False
+                self.handle_transfer_event(event)
+
+    def handle_transfer_event(self, event: dict):
+        if event['event'] != 'Transfer':
+            logger.warning('Failed to parse non-transfer event')
+            return
+        key_to = self.keys['to']
+        key_from = self.keys['from']
+        key_amount = self.keys['amount']
+        block = event['blockNumber']
+        if event['args'][key_from].lower() in self.addresses.values():
+            self.update_balance(event['args'][key_from].lower(),
+                                -1 * event['args'][key_amount], block)
+        if event['args'][key_to].lower() in self.addresses.values():
+            self.update_balance(event['args'][key_to].lower(),
+                                event['args'][key_amount], block)
+
+    def update_balance(self, account: str, change: int, block: int):
+        last_updated_block = self.last_update[account]
+        self.balances[account][block] = self.balances[account][last_updated_block] + change
+        self.last_update[account] = block
+
+    def set_quantity_key(self, event: dict) -> bool:
+        """ERC20 contracts don't follow a standard name for transfer args.
+        This should be handled here."""
+        keys = event['args'].keys()
+        if '_from' in keys:
+            self.keys['from'] = '_from'
+        elif 'from' not in keys:
+            logger.error('Transfer arg: `from` not found in event parser')
+            return False
```

If this fails it's pointless to run so I would rather raise an error here.

---

True. That goes for the entire function. If any unknown key is found, it'll raise an error.